src/jit/lowerarmarch.cpp

```diff
@@ -280,93 +253,68 @@ void Lowering::LowerBlockStore(GenTreeBlk* blkNode)
             // it to a larger constant whose size is sufficient to support
             // the largest width store of the desired inline expansion.
 
-            ssize_t fill = initVal->gtIntCon.gtIconVal & 0xFF;
+            ssize_t fill = src->AsIntCon()->IconValue() & 0xFF;
+
             if (fill == 0)
             {
-                MakeSrcContained(blkNode, source);
+                src->SetContained();
             }
-            else if (size < REGSIZE_BYTES)
+            else if (size >= REGSIZE_BYTES)
             {
-                initVal->gtIntCon.gtIconVal = 0x01010101 * fill;
+                fill *= 0x0101010101010101LL;
+                src->gtType = TYP_LONG;
             }
             else
             {
-                initVal->gtIntCon.gtIconVal = 0x0101010101010101LL * fill;
-                initVal->gtType             = TYP_LONG;
+                fill *= 0x01010101;
             }
-            blkNode->gtBlkOpKind = GenTreeBlk::BlkOpKindUnroll;
+
+            src->AsIntCon()->SetIconValue(fill);
         }
         else
 #endif // _TARGET_ARM64_
         {
+            // TODO-ARM-CQ: Currently we generate a helper call for every initblk we encounter.
+            // Later on we should implement loop unrolling code sequences to improve CQ.
             blkNode->gtBlkOpKind = GenTreeBlk::BlkOpKindHelper;
         }
     }
     else
     {
-        // CopyObj or CopyBlk
-        // Sources are src and dest and size if not constant.
+        assert(src->OperIs(GT_IND, GT_LCL_VAR, GT_LCL_FLD));
+        src->SetContained();
 
-        if (blkNode->OperGet() == GT_STORE_OBJ)
+        if (src->OperIs(GT_IND))
         {
-            // CopyObj
-            GenTreeObj* objNode = blkNode->AsObj();
-
-            unsigned slots = objNode->GetLayout()->GetSlotCount();
-
-#ifdef DEBUG
-            // CpObj must always have at least one GC-Pointer as a member.
-            assert(objNode->GetLayout()->HasGCPtr());
-
-            assert(dstAddr->gtType == TYP_BYREF || dstAddr->gtType == TYP_I_IMPL);
+            // Make sure that GT_IND lowering didn't mark the source address as contained,
+            // containment is supported only in certain cases.
```

Comment:
It is not clear why `ClearContained` is called. It is not a typical thing to do in lowering, it's usually done only when node lowering changes the IR in a way that may invalidate containment checks done previously. It's likely that this should just be an assert, `ContainCheckIndir` does reject `TYP_STRUCT` indirs so I don't believe that the address can ever be contained at this point. I suppose it would be good to add an assert but also keep `ClearContained` around for safety, for now.

Local address lowering is still done, later in the code. We have:
* containing the IND is required
* not containing the address is always correct
* the address containment logic will become a bit more complex and it is always optional

So the first 2 are more like preconditions and it makes more sense to put them at the top of the block while leaving the 3rd at the end of the block (and maybe moved to a separate `CheckContainBlockStore` function)

Reply:
> Since Lowering operates bottom-up, it can mark a GT_IND child as contained

Yes but it doesn't do that because these IND nodes have TYP_STRUCT type: https://github.com/dotnet/coreclr/blob/088299e2b7f439dfd1f5056fdc291b9d82149fc1/src/jit/lowerxarch.cpp#L1554-L1561

Which makes sense, such a node always has to be contained so its user should deal with it, including the address.

Reply:
I see - so perhaps it should indeed be an error/assert?

Reply:
Yes, I've added an assert. 

I didn't remove `ClearContained` yet, just in case. I'm a bit concerned about how containment works, we assume that nodes are not contained by default but in some cases lowering changes may require revisiting previous containment decisions. Though that's probably impossible in this case thanks to the rather specific shape of block ops, maybe I'll convince myself that this is useless and remove it in a later PR :)